### ISSUE TITLE
Is UploadMediaAsync what Twitter considers to be 'async'?

### ISSUE THREAD

**Issue by Kew** [JoeMayo/LinqToTwitter#70]
I am trying to upload videos greater than 30 seconds and 15M by using async upload (As per https://dev.twitter.com/rest/media/uploading-media )

> Duration should be between 0.5 seconds and 30 seconds (sync) / 140 seconds (async)
> File size should not exceed 15 mb (sync) / 512 mb (async)

Issue is that twitter seems to behave as if I was using 'sync' and reports that files over 30 seconds or 15M that I attempt to upload as too long or too big.

Here is my code, can please you see why Twitter might see this request as not being 'async' ?

```
using(var twitterCtx = new LinqToTwitter.TwitterContext(postAuth)) 
{
        var uploadedMedia = await twitterCtx.UploadMediaAsync(File.ReadAllBytes(videoPath), "video/mp4");           

        var mediaIds = new List<ulong> { uploadedMedia.MediaID };

        Status tweet = await twitterCtx.TweetAsync (
            message,
            mediaIds
        );
}
```

**Comment by JoeMayo** [comment]
Hi Kew,

Are videos less than 30 seconds and smaller that 15 Mb working for you? What is the exact size and duration you're using?

Older versions of LINQ to Twitter used the sync endpoint because Twitter only supported images at that time. However, the latest version of LINQ to Twitter supports the async endpoint exclusively. Are you using the latest version?

If so, please share some exception data. The TwitterQueryException has StatusCode, ErrorCode, and ReasonPhrase, as well as Stack Trace.

You might try the Upload Video demo in the Console Samples in the source code to see if that's working for you. 

Also, check videoPath to make sure it's the same mime type that you're passing in the 2nd parameter.

[@JoeMayo](https://twitter.com/JoeMayo)

**Comment by Kew** [comment]
Yes uploading less than 15 MB is working for me. I am using version 4.1.0 and compiling mono using Xamarin. I'll get a dump from the exception to you in the next few days

**Comment by Kew** [comment]
Here is the exception with requested fields.

```
StatusCode = 'BadRequest'
ReasonPhrase = 'Bad Request'
ErrorCode = '324'
==========================================================================================================================
InnerExceptions: LinqToTwitter.TwitterQueryException: Duration too long, maximum:30000, actual:57520 (MediaId: snf:751972440478130177)
   at LinqToTwitter.Net.TwitterErrorHandler.BuildAndThrowTwitterQueryException(String responseStr, HttpResponseMessage msg)
   at LinqToTwitter.Net.TwitterErrorHandler.<HandleGenericErrorAsync>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at LinqToTwitter.Net.TwitterErrorHandler.<ThrowIfErrorAsync>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at LinqToTwitter.TwitterExecute.<HandleResponseAsync>d__59.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at LinqToTwitter.TwitterExecute.<PostToTwitterAsync>d__58`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at LinqToTwitter.TwitterContext.<TweetOrReplyAsync>d__211.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at LinqToTwitter.TwitterContext.<TweetAsync>d__209.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at UnitySocial.UnitySocial.<Twitter>d__10.MoveNext() in 
```

**Comment by Kew** [comment]
Here is a link to a working tweet from Linq2Twitter.

https://twitter.com/qewbe/status/750549551585562624

Thanks!

**Comment by Kew** [comment]
Compiled the Linq2Twitter tests from master on GitHub.

Was able to upload your test video.

But when I used the inbuilt console tests to upload a 57 second video I got the same error that I am getting with my program.

![bandicam 2016-07-10 14-10-27-869](https://cloud.githubusercontent.com/assets/1836917/16711615/67d50d34-46a8-11e6-9e9d-ed42134f5d5e.png)

**Comment by JoeMayo** [comment]
Kew,

Thank you. This is helpful.

Joe

**Comment by DrMiaow** [comment]
+1

This is blocking me too. :(

**Comment by JoeMayo** [comment]
I just checked in a fix for this. For longer videos, there's a new video_category parameter, videoCategory in LINQ to Twitter. You can use it like this:

        Media media = await twitterCtx.UploadMediaAsync(
            File.ReadAllBytes(@"..\..\images\SampleVideo.mp4"), "video/mp4", "tweet_video");

Notice the "tweet_video" parameter. This would be tweet_image for pictures and tweet_gif for GIF animations.

I have a few more items to add before pushing a new NuGet package, but the source code is available now.

[@JoeMayo](https://twitter.com/JoeMayo)

**Comment by sqlcnet** [comment]
This is GREAT news Joe. Thank you, thank you, thank you! No more banging my head against the wall. :) I’m looking forward to the NuGet.

**Comment by DrMiaow** [comment]
Thankyou, I will test this out ASAP and get back to you with feedback.

**Comment by Kew** [comment]
Thank you, I will try it out.

**Comment by Muath89** [comment]
+1

I have tried to use 'media_category' as in 
Media media = await twitterCtx.UploadMediaAsync(
        File.ReadAllBytes(@"..\..\images\SampleVideo.mp4"), "video/mp4", "tweet_video");
but the UploadMediaAsync function has no overloaded method matched. 
I am using version v4.1.0

**Comment by JoeMayo** [comment]
@Muath89 The changes are checked into code, but aren't on NuGet yet. I typically batch updates before pushing to NuGet. That also gives early adopters a chance to look at the work I've done and receive feedback before releasing to the rest of the world.

[@JoeMayo](https://twitter.com/JoeMayo)

**Comment by sqlcnet** [comment]
With your new fix in GitHub, I have been able to tweet videos more than 30 secs in duration and that were less than 15MB in size. However, when trying to tweet a video that is more than 30 seconds AND greater than 15MB in size it fails with:

Message: Not valid video 
ReasonPhrase: Bad Request 
ErrorCode: 324

So Joe, the 64,000 dollar question is... during your unit testing were you able to successfully tweet a video > 30 sec that was > 15MB? If so, then perhaps some of the other criteria is incorrect, such as my frame rate, etc.

It really would be good to know if YOU were successful with the > 15MB videos. 

Thanks Joe.

**Comment by Muath89** [comment]
@JoeMayo , Thank you.
I have tried the new code but it doesn't work with large video size. Errro: "Not valid video".
And kindly when do you expect to push your changes to NuGet?

**Comment by JoeMayo** [comment]
I looked at the video I tested with and noticed that it was > 30 seconds, but < 15MB. I'll go back and look at this again to find out what is wrong.

**Comment by DrMiaow** [comment]
Same :( But you seem to be onto it :)

**Comment by sqlcnet** [comment]
@JoeMayo 

Hey Joe, this may be helpful. See this complete Python example of an upload and tweet of a large video > 30 secs and > 15 MB    -->   https://github.com/twitterdev/large-video-upload-python

Points of interest to me:
1. It uses 'tweetvideo' instead of 'tweet_video' as the media category. I tried it and this alone did not solve the problem though in your latest github code.
2. It uses 4 MB chunks instead of your 5 MB. Not sure this causes a problem but I have read about others indicating 5 MB did not work and they decreased it.
3. And mostly, it uses the STATUS call to repeatedly check status every 5 seconds to ensure the media processing has completed before using the media id in a tweet. Perhaps this is twitter needing to run the video thru a conversion first before the media id is valid. Perhaps you already do the STATUS call, but I'm kinda thinking no. 

I hope this helps.

**Comment by Kew** [comment]
My videos are less than 15M but 2 minutes and 20 seconds, which is twitters new limit of 140 seconds.

`var uploadedMedia = await twitterCtx.UploadMediaAsync(File.ReadAllBytes(videoPath), "video/mp4", "tweet_video");			`

I get this with your latest DLL.


```
==========================================================================================================================
InnerExceptions: LinqToTwitter.TwitterQueryException: Not valid video
   at LinqToTwitter.Net.TwitterErrorHandler.BuildAndThrowTwitterQueryException(String responseStr, HttpResponseMessage msg)
   at LinqToTwitter.Net.TwitterErrorHandler.<HandleGenericErrorAsync>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at LinqToTwitter.Net.TwitterErrorHandler.<ThrowIfErrorAsync>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at LinqToTwitter.TwitterExecute.<HandleResponseAsync>d__59.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at LinqToTwitter.TwitterExecute.<PostToTwitterAsync>d__58`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at LinqToTwitter.TwitterContext.<TweetOrReplyAsync>d__211.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at LinqToTwitter.TwitterContext.<TweetAsync>d__209.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at UnitySocial.UnitySocial.<Twitter>d__10.MoveNext() in C:\Users\Kew\Documents\GitHub\Rivalry\UnitySocial\UnitySocial\UnitySocial.cs:line 622
==========================================================================================================================
StatusCode   = 'BadRequest'
ReasonPhrase = 'Bad Request'
ErrorCode    = '324'
```

**Comment by JoeMayo** [comment]
I think I got it this time. I pushed the latest changes, based on @sqlcnet comments (thanks), that includes media status queries. Here's an example, which you can find in the ConsoleSamples/MediaDemos.cs:

```C#
        static async Task UploadVideoAsync(TwitterContext twitterCtx)
        {
            string status =
                "Testing video upload tweet #Linq2Twitter £ " +
                DateTime.Now.ToString(CultureInfo.InvariantCulture);

            //Media media = await twitterCtx.UploadMediaAsync(File.ReadAllBytes(@"..\..\images\LinqToTwitterNormalTest.mp4"), "video/mp4", "tweet_video");
            Media media = await twitterCtx.UploadMediaAsync(File.ReadAllBytes(@"..\..\images\LinqToTwitterMediumTest.mp4"), "video/mp4", "tweet_video");
            //Media media = await twitterCtx.UploadMediaAsync(File.ReadAllBytes(@"..\..\images\LinqToTwitterErrorTest.mp4"), "video/mp4", "tweet_video");

            Media mediaStatusResponse = null;
            do
            {
                if (mediaStatusResponse != null)
                {
                    int checkAfterSeconds = mediaStatusResponse?.ProcessingInfo?.CheckAfterSeconds ?? 0;
                    Console.WriteLine($"Twitter video testing in progress - waiting {checkAfterSeconds} seconds.");
                    await Task.Delay(checkAfterSeconds * 1000);
                }

                mediaStatusResponse =
                    await
                    (from stat in twitterCtx.Media
                     where stat.Type == MediaType.Status &&
                           stat.MediaID == media.MediaID
                     select stat)
                    .SingleOrDefaultAsync(); 
            } while (mediaStatusResponse?.ProcessingInfo?.State == MediaProcessingInfo.InProgress);

            if (mediaStatusResponse?.ProcessingInfo?.State == MediaProcessingInfo.Succeeded)
            {
                Status tweet = await twitterCtx.TweetAsync(status, new ulong[] { media.MediaID });

                if (tweet != null)
                    Console.WriteLine($"Tweet sent: {tweet.Text}");
            }
            else
            {
                MediaError error = mediaStatusResponse?.ProcessingInfo?.Error;

                if (error != null)
                    Console.WriteLine($"Request failed - Code: {error.Code}, Name: {error.Name}, Message: {error.Message}");
            }
        }
```
The source code also has a LinqToTwitterErrorsTest.mp4 to help test error handling. Please let me know if you have any more problems.

**Comment by sqlcnet** [comment]
Hey Joe. I don't see any new changes in the master branch. The last changes show from 15 days ago.  I'm new to GitHub. Is there somewhere else I should look?

**Comment by JoeMayo** [comment]
I just pushed the changes and you should see them now.

**Comment by DrMiaow** [comment]
Thanks, will try this tonight.

**Comment by sqlcnet** [comment]
Success! These latest changes worked perfectly for me. Thanks Joe! I'll wait to move my code to production until you have this fix officially on NuGet. Please let us know when that occurs.

**Comment by Kew** [comment]
Trying to get this into a beta release of my game.

When I build the latest version and copy in the LinqToTwitter.net.dll and LinqToTwitter.AspNet.dll over the top of the ones I get from NuGet I get the following error.

`Error: Method not found: System.Threading.Tasks.Task 1<LinqToTwitter.Media>LinqToTwitter.TwitterContext.UploadMediaAsync(Byte[], System.String, System.Threading.CancellationToken).`

Any idea what I am doing wrong?

**Comment by sqlcnet** [comment]
Not sure... but know there is a new additional parm in his latest code that is required... media_category (for "tweet_video"). For example: 

`var media = await twitterContext.UploadMediaAsync(data, "video/mp4",  "tweet_video", cancelToken);
`

There are 4 parms now instead of 3. Cancel Token is optional.

**Comment by Kew** [comment]
Thanks. That was it. I'm just linking directly to the DLLs and using @JoeMayo's status checking code above, and it works! 

https://twitter.com/qewbe/status/773797103717842944  :)

Once the Nuget package comes out I will revert the changes to my project.

Thankyou!

**Comment by DrMiaow** [comment]
This works! 

Thankyou!

**Comment by nikunjce** [comment]
can you please upgrade the Nuget package so we can easily use it in our project.
I have directly reference the dlls but am getting "Not a valid" video error.

**Comment by JoeMayo** [comment]
I just pushed an update to NuGet as a beta release: https://www.nuget.org/packages/linqtotwitter/4.2.0-beta1

**Comment by nikunjce** [comment]
Thanks

**Comment by JoeMayo** [comment]
This was fixed in [LINQ to Twitter v4.2.0](https://github.com/JoeMayo/LinqToTwitter/releases/tag/v4.2.0).

[@JoeMayo](https://twitter.com/JoeMayo)
